File: programmers/베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

struct genre {
    int total;
    vector<pair<int, int> > play_list;
};

bool cmp(pair<int, int> a, pair<int, int> b) {
    if (a.first == b.first) return a.second < b.second;
    return a.first > b.first;
}

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    map<string, genre> m;
    map<int, vector<pair<int, int> >, greater<int> > res;
    
    int length = genres.size();
    for (int index = 0; index < length; index++) {
        m[genres[index]].total += plays[index];
        m[genres[index]].play_list.push_back({plays[index], index});
    }
    
    for (auto items : m) {
        sort(items.second.play_list.begin(), items.second.play_list.end(), cmp);
        res.insert({items.second.total, items.second.play_list});
    }
    
    for (auto result : res) {
        int count = 0;
        for (auto times : result.second) {
            if (count >= 2) break;
            answer.push_back(times.second);
            count++;
        }
    }
    return answer;
}

File: programmers/베스트앨범/베스트앨범.kt
import java.util.*
class Solution {
    fun solution(genres: Array<String>, plays: IntArray): IntArray {
        val dataSize = genres.size
        val genreList = Array(dataSize) { mutableListOf<Pair<Int, Int>>() }
        var genreCount = 0
        val genre = TreeMap<String, Int>()
        
        repeat(dataSize) { index ->
            var myIndex: Int
            when (genre.contains(genres[index])) {
                true -> myIndex = genre[genres[index]]!!
                false -> {
                    myIndex = genreCount++
                    genre[genres[index]] = myIndex
                }
            }
            genreList[myIndex].add(Pair(plays[index], index))
        }
        
        val genreSort = mutableListOf<Pair<Int, Int>>()
        for (g in genre) {
            var mySum = 0
            genreList[g.value].sortByDescending { it.first }
            
            for (play in genreList[g.value]) {
                mySum += play.first
            }
            genreSort.add(Pair(mySum, g.value))
        }
        
        genreSort.sortByDescending { it.first }
        var answer = mutableListOf<Int>()
        for (g in genreSort) {
            genreList[g.second].forEachIndexed { index, g -> 
                if (index < 2) {
                    answer.add(g.second)
                }
            }
        }
        return answer.toIntArray()
    }
}